### DISCUSSION TITLE
New greenfield event: InvoicePaymentSettled

### DISCUSSION THREAD

**Discussion by woutersamaey** [btcpayserver/btcpayserver#2691] [Technical Support]
@Kukks  and I discussed the need for a new event "Payment Settled". This event fills the hole where a payment is late and the "Invoice Settled" event cannot be used anymore.

This new event should be the same as "Payment Received" except it waits until settlement (enough confirmations).

Merchants who don't care about exchange rates (one's who are invoicing in BTC and not fiat) may not care about expiry or fiat exchange rates so if a payment comes in 2 days late and they're fine with that, they should be able to use this event to track for late payments and act accordingly.

This idea came from a merchant who had this requirement and was posted in #2671

**Comment by bolatovumar** [comment]
`PaymentReceived` only exists for payment requests, not as a webhook event type. Do you mean we should add a `PaymentSettled` webhook event type which would be similar to how `PaymentReceived` on payment requests works?

![Screen Shot 2021-07-09 at 12 08 00 PM](https://user-images.githubusercontent.com/1934678/125125468-7d891800-e0ae-11eb-977d-e938a49d8c9b.png)

**Reply by woutersamaey** [reply]
We want a webhook for each and every payment. So not related to Payment Requests, but invoices as well.
The point is that we want to start handling late payments and that's currently not possible once an invoice is expired.

These 2 events should fire on any payment, regardless of timing.
I know there are limits, but the Invoice Expiry should not be the last of the webhooks.

Actually, I do see it in your screenshot, but it's named `InvoicePaymentReceived` not `PaymentReceived`. So the new event should be `InvoicePaymentSettled` actually.

**Reply by bolatovumar** [reply]
@woutersamaey `InvoicePaymentReceived` doesn't exist anywhere: https://github.com/btcpayserver/btcpayserver/search?q=InvoicePaymentReceived

**Reply by woutersamaey** [reply]
Sorry it is `InvoiceReceivedPayment`, the 2nd one in your screenshot!

**Reply by bolatovumar** [reply]
@woutersamaey ok, so this new event name would be `InvoiceSettledPayment` then.

**Reply by woutersamaey** [reply]
I guess this is where it gets confusing. 

The idea here is that we have 2 kinds of events. We do invoice events that respond to invoice changes. And we do (separately) Payment events that notify of payments. 

For the Payment events, we see 2 events: the one we already have on 0 confirmations and the new one once there are sufficient confirmations and call that PaymentSettled.

I didn't realize that in the code the word "Invoice" is there, because it's not called that in the docs. This extra word adds to confusion. 

It suggest we look at it like a different entity (a payment) and that has events, just like you could have Store or User events (i.e. on create, update or delete).
